File: ED/solucionRel4/Relación ejercicios java/src/dataStructures/heap/MaxiphobicHeap.java
/**
 * @author <<write here your name>> 
 * 
 * Maxiphobic Heaps
 * Data Structures, Grado en Informática. UMA.
 *
 */
 
package dataStructures.heap;


/**
 * Heap implemented using maxiphobic heap-ordered binary trees.
 * @param <T> Type of elements in heap.
 */
public class MaxiphobicHeap<T extends Comparable<? super T>> implements	Heap<T> {

	protected static class Tree<E> {
		E elem;
		int size;
		int weight;
		Tree<E> left;
		Tree<E> right;
	}

	public int size() {
		return size(root);
	}

	private static int size(Tree<?> heap) {
		return heap == null ? 0 : heap.size;
	}

	private static int weight(Tree<?> heap) {
		return heap == null ? 0 : heap.weight;
	}

	private static <T extends Comparable<? super T>> Tree<T> merge(Tree<T> h1,	Tree<T> h2) {
		// TODO
		if (h1==null) {
			return h1;
		}

		if (h2==null) {
			return h2;
		}

		if (h2.elem.compareTo(h1.elem)<0) {
			Tree<T> aux = h1;
			h1=h2;
			h2=aux;
		}

		if (h2.size>=h1.left.size && h2.size>=h1.right.size) {
			h1.right = merge(h1.left,h1.right);
			h1.left = h2;
		} else if (h1.left.size>=h1.right.size && h1.left.size>=h2.size) {
			h1.right=merge (h1.right,h2);
		} else {
			Tree<T> aux = h1.right;
			h1.right = merge(h1.left,h2);
			h1.left = aux;
		}

		return h1;
	}

	protected Tree<T> root;

	/**
	 * Constructor: creates an empty Maxiphobic Heap.
	 * <p>Time complexity: ???
	 */
	public MaxiphobicHeap() {
		// TODO
		root = null;
	}

	/**
	 * <p>Time complexity: ???
	 */
	public boolean isEmpty() {
		// TODO
		return this == null;
	}

	/**
	 * <p>Time complexity: ???
	 * @throws <code>EmptyHeapException</code> if heap stores no element.
	 */
	public T minElem() {
		// TODO
		return this.root.elem;
	}

	/**
	 * <p>Time complexity: ???
	 * @throws <code>EmptyHeapException</code> if heap stores no element.
	 */
	public void delMin() {
		// TODO
		root = merge(root.left,root.right);
	}

	/**
	 * <p>Time complexity: ???
	 */
	public void insert(T value) {
		// TODO
		Tree<T> single = new Tree<>();
		single.elem = value;
		single.size=1;
		single.left=null;
		single.right=null;
		root = merge(root,single);
	}

	private static String toStringRec(Tree<?> tree) {
		return tree == null ? "null" : "Node<" + toStringRec(tree.left) + ","
				+ tree.elem + "," + toStringRec(tree.right) + ">";
	}

	/**
	 * Returns representation of heap as a String.
	 */
	public String toString() {
		String className = getClass().getSimpleName();

		return className+"("+toStringRec(this.root)+")";
	}

}

File: ED/solucionRel4/haskell/DataStructures/Práctica tema 4 heap/MaxiphobicHeap.hs
-------------------------------------------------------------------------------
-- Student's name: Javier Ortuño Roig
--
-- Maxiphobic Heaps
-- Data Structures. Grado en Informática. UMA.
-------------------------------------------------------------------------------

module DataStructures.Heap.MaxiphobicHeap
  ( Heap
  , empty
  , isEmpty
  , minElem
  , delMin
  , insert
  , merge
  , mkHeap
  ) where

import Test.QuickCheck

data Heap a  = Empty | Node a Int (Heap a) (Heap a) deriving Show

-- number of elements in tree rooted at node
weight :: Heap a -> Int
weight Empty           = 0
weight (Node _ w _ _)  = w


singleton :: a -> Heap a
singleton x  = Node x 1 Empty Empty


empty :: Heap a
empty = Empty


isEmpty :: Heap a -> Bool
isEmpty Empty = True
isEmpty _ = False


insert :: (Ord a) => a -> Heap a -> Heap a
insert x h = merge (singleton x) h


minElem :: Heap a -> a
minElem Empty = error "monticulo vacio"
minElem (Node x p a b) = x


delMin :: (Ord a) => Heap a -> Heap a
delMin (Node a w hl hr) = merge hl hr


merge :: (Ord a) => Heap a -> Heap a -> Heap a -- T(n) = log (3/2) n
merge Empty Empty = Empty
merge Empty b = b
merge a Empty = a
merge h@(Node x w hl hr) h'@(Node x' w' hl' hr')
    | x <= x' = (Node x ws h1 (merge h2 h3))
    | otherwise = (Node x' ws h1' (merge h2' h3'))
    where
      sortHeaps :: Heap a -> Heap a -> Heap a -> (Heap a, Heap a, Heap a)
      sortHeaps h1 h2 h3
        | w1 >= w2 && w1 >= w3 = (h1,h2,h3)
        | w2 >= w3 && w2 >= w1 = (h2,h1,h3)
        | otherwise = (h3,h1,h2)
          where 
            w1 = weight h1
            w2 = weight h2
            w3 = weight h3
        
      ws = w + w'
      (h1,h2,h3) = sortHeaps hl hr h' 
      (h1',h2',h3') = sortHeaps hl' hr' h

{-Otra forma  
merge :: (Ord a) => Heap a -> Heap a -> Heap a -- T(n) = log (3/2) n
merge Empty Empty = Empty
merge Empty b = b
merge a Empty = a
merge h@(Node a w hl hr) h'@(Node c w' hl' hr')
    | a <= c && weight hl >= weight hr && weight hl >= w' = (Node a (weight hr + w') hl (merge hr h'))
    | a <= c && weight hr >= weight hl && weight hr >= w' =  (Node a (w + weight hl) hr (merge h hl))
    | a <= c && w' >= weight hr && w' >= weight hl = (Node a (weight hr + weight hl) h' (merge hr hl))
    | a > c && weight hl >= weight hr && weight hl >= w = (Node c (w + weight hr) hl (merge hr h))
    | a > c && weight hr >= weight hl && weight hr >= w =  (Node c (w + weight hl) hr (merge h hl))
    | a > c && w >= weight hr && w >= weight hl = (Node c (weight hr + weight hl) h (merge hr hl))
-}

-- Efficient O(n) bottom-up construction for heaps
mkHeap :: (Ord a) => [a] -> Heap a
mkHeap []  = empty
mkHeap xs  = mergeLoop (map singleton xs)
  where
    mergeLoop [h]  = h
    mergeLoop hs   = mergeLoop (mergePairs hs)

    mergePairs []         = []
    mergePairs [h]        = [h]
    mergePairs (h:h':hs)  = merge h h' : mergePairs hs

-------------------------------------------------------------------------------
-- Generating arbritray Heaps
-------------------------------------------------------------------------------

instance (Ord a, Arbitrary a) => Arbitrary (Heap a) where
  arbitrary  = do
    xs <- arbitrary
    return (mkHeap xs)
